addons/binding/org.openhab.binding.opendaikin/src/main/java/org/openhab/binding/opendaikin/internal/handler/OpenDaikinAcUnitHandler.java

```diff
@@ -0,0 +1,267 @@
+        }
+
+        SensorInfo sensorInfo = webTargets.getSensorInfo();
+        if (sensorInfo != null) {
+            if (sensorInfo.indoortemp.isPresent()) {
+                updateTempState(OpenDaikinBindingConstants.CHANNEL_INDOOR_TEMP, sensorInfo.indoortemp.get());
+            } else {
+                updateState(OpenDaikinBindingConstants.CHANNEL_INDOOR_TEMP, UnDefType.UNDEF);
+            }
+
+            if (sensorInfo.outdoortemp.isPresent()) {
+                updateTempState(OpenDaikinBindingConstants.CHANNEL_OUTDOOR_TEMP, sensorInfo.outdoortemp.get());
+            } else {
+                updateState(OpenDaikinBindingConstants.CHANNEL_OUTDOOR_TEMP, UnDefType.UNDEF);
+            }
+
+            if (sensorInfo.indoorhumidity.isPresent()) {
+                updateState(OpenDaikinBindingConstants.CHANNEL_HUMIDITY,
+                        new DecimalType(sensorInfo.indoorhumidity.get()));
+            } else {
+                updateState(OpenDaikinBindingConstants.CHANNEL_HUMIDITY, UnDefType.UNDEF);
+            }
+        }
+    }
+
+    private void updateTempState(String channel, double temp) {
+        QuantityType<Temperature> tempToUpdateWith = new QuantityType<Temperature>(new DecimalType(temp),
+                SIUnits.CELSIUS);
+
+        updateState(channel, tempToUpdateWith);
+    }
+
+    private void changePower(boolean power) throws OpenDaikinCommunicationException {
+        ControlInfo info = webTargets.getControlInfo();
+        info.power = power;
+        webTargets.setControlInfo(info);
+    }
+
+    private void changeSetPoint(double temp) throws OpenDaikinCommunicationException {
+        ControlInfo info = webTargets.getControlInfo();
+
+        if (useFahrenheitForChannel(OpenDaikinBindingConstants.CHANNEL_AC_TEMP)) {
+            info.temp = Optional.of(fToC(temp));
+        } else {
+            info.temp = Optional.of(temp);
+        }
+
+        webTargets.setControlInfo(info);
+    }
+
+    private boolean useFahrenheitForChannel(String channel) {
+        return this.getThing().getChannel(channel) != null && Boolean.TRUE.equals(this.getThing().getChannel(channel)
+                .getConfiguration().get(OpenDaikinBindingConstants.SETTING_USE_FAHRENHEIT));
+    }
+
+    private void changeMode(ControlInfo.Mode mode) throws OpenDaikinCommunicationException {
+        ControlInfo info = webTargets.getControlInfo();
+        info.mode = mode;
+        webTargets.setControlInfo(info);
+    }
+
+    private void changeFanSpeed(ControlInfo.FanSpeed fanSpeed) throws OpenDaikinCommunicationException {
+        ControlInfo info = webTargets.getControlInfo();
+        info.fanSpeed = fanSpeed;
+        webTargets.setControlInfo(info);
+    }
+
+    private void changeFanDir(ControlInfo.FanMovement fanDir) throws OpenDaikinCommunicationException {
+        ControlInfo info = webTargets.getControlInfo();
+        info.fanMovement = fanDir;
```

Comment:
This looks badly encapsulated, please consider using setters or maybe even some builder like structure?

Do I read correctly that in order to change the direction you first call the endpoint update the response you get and then send a request to change? That feels non-optimal to me? 

It could (and in my experience) will lead to concurrency problems.

Reply:
This is definitely not a great pattern, but is due to a design decision (flaw) in the way Daikin set up their API. You must pass the A/C units a full structure with all of the fields filled out to make any change. As a result, it's not actually possible to just increase the temperature one degree or turn on fan movement. To simulate making a single action I read the current state of the A/C unit and then make the single change required and pass that back.